libs/community/langchain_community/document_loaders/parsers/pdf.py

```diff
@@ -1464,6 +1471,326 @@ def _extract_images_from_page(self, page: pdfplumber.page.Page) -> str:
+        if not images_parser:
+            images_parser = RapidOCRBlobParser()
+        self.mode = mode
+        self.pages_delimiter = pages_delimiter
+        self.images_parser = images_parser
+        self.images_inner_format = images_inner_format
+        self.extract_tables = extract_tables
+
+        self.cleanup = cleanup
+        self.concurrency = concurrency
+        self.maintain_format = maintain_format
+        self.model = model
+        if not custom_system_prompt:
+            custom_system_prompt = ZeroxPDFParser._prompt
+        self.custom_system_prompt = custom_system_prompt
+        self.select_pages = select_pages
+        self.zerox_kwargs = zerox_kwargs
+
+    @staticmethod
+    def _is_valid_url(url: str) -> bool:
+        """Check if the url is valid."""
+        parsed = urlparse(url)
+        return bool(parsed.netloc) and bool(parsed.scheme)
+
+    def lazy_parse(self, blob: Blob) -> Iterator[Document]:  # type: ignore[valid-type]
+        """Lazily parse the blob.
+
+        Args:
+            blob: The blob to parse.
+
+        Raises:
+            ImportError: If the `py-zerox` package is not installed.
+
+        Yields:
+            An iterator over the parsed documents.
+        """
+        try:
+            from pyzerox import zerox
+        except ImportError:
+            raise ImportError(
+                "Could not import pyzerox python package. "
+                "Please install it with `pip install py-zerox`."
+            )
+        temp_file = None
+        try:
+            if not ZeroxPDFParser._is_valid_url(str(blob.path)):
+                temp_file = NamedTemporaryFile()
+                with open(temp_file.name, "wb") as f:
+                    f.write(blob.as_bytes())
+                file_path = temp_file.name
+            else:
+                file_path = str(blob.path)
+
+            with blob.as_bytes_io() as pdf_file_obj:
+                doc_metadata = _purge_metadata(self._get_metadata(pdf_file_obj))
+
+            doc_metadata["source"] = blob.source or blob.path
+            zerox_prompt = self.custom_system_prompt
+
+            if not zerox_prompt and self.images_parser or self.extract_tables:
+                prompt_tables = ZeroxPDFParser._map_extract_tables[self.extract_tables]
+                clazz = self.images_parser.__class__
+                if clazz in ZeroxPDFParser._map_extract_images:
+                    prompt_images = ZeroxPDFParser._map_extract_images[clazz]
+                else:
+                    if not ZeroxPDFParser._warn_creator:
+                        ZeroxPDFParser._warn_creator = True
+                        logger.warning("images_parser can not be simulated")
+                    prompt_images = ""
+                zerox_prompt = PromptTemplate.from_template(
+                    self.custom_system_prompt
+                ).format(prompt_tables=prompt_tables, prompt_images=prompt_images)
+            zerox_output = asyncio.run(
```

Comment:
This needs to be kicked off on a thread? there might be an existing even loop

Reply:
This is the current code validated in the main branch.
I'll see what I can do.

Reply:
It's a big problem, with a lot of impact. There is no asynchronous alternative to `lazy_parse()`.
If there were an `alazy_parse()` it would be cleaner. Zerox only works in `async`.
Where can I find the thread to discuss this?